feat: Adds encryption backend config and support for PKCS11 backend (#253)
canonical/notary: docs/how-to/hsm.md

## Hardware Security Modules (HSMs)

Prerequisites:
- An HSM that supports the PKCS11 protocol
- AES256 symmetric key created on the HSM with capabilities to encrypt and decrypt using the AES-CBC algorithm
- Access to the HSM's driver/interface library file (.so or .dylib file) installed with the HSM's SDK
- The HSM's connector up and running

Configuration
1. Provide a name to your backend, in the following example we call our backend yubihsm2-backend
2. Provide the path to the library that is installed with the SDK of your HSM
3. Provide the Pin to login on your HSM, this will be in the following format: <auth key id><password>, so if the authentication key used has the id 0001 (default key) your pin might look something like "0001password"
4. Provide the ID of the symmetric encryption key that will be used

```yaml
encryption_backend:
  yubihsm2-backend: # name of the backend
    pkcs11:
      lib_path: "/usr/lib/x86_64-linux-gnu/pkcs11/yubihsm_pkcs11.so"
      pin: "0001password"
      key_id: 0x1234

> ```

Reviewer: What do you mean by " we can't specify a config file for the snap"? The config file is available after installing the snap under `/var/snap/notary/common/notary.yaml` and that's where users can change it prior to starting notary (`sudo snap start notary.notaryd`)

> ```

Author: I made them 2 points and the hsm related config is under 2.

# Use an HSM as Notary's Encryption Backend

> In this guide we show the required configuration to use a Hardware Security Module (HSM) as an encryption backend for Notary.

Reviewer: ```suggestion
In this guide we walk you through the required steps to configure and use a Hardware Security Module (HSM) as an encryption backend for Notary.
```

**Note**: 
- Once Notary is initialized it must continue using the encryption backend configured at the time of initialization, at the moment there is no way to switch backends.
- This has been tested with YubiHSM2, while this should work with any HSM that supports the PKCS11 protocol please follow any further instruction from the specific HSM vendor.

> ## Hardware Security Modules (HSMs)

Reviewer: ```suggestion
```

Author: :white_check_mark:

### Prerequisites:
- An HSM that supports the PKCS11 protocol
- AES256 symmetric key created on the HSM with capabilities to encrypt and decrypt using the AES-CBC algorithm
- Access to the HSM's driver/interface library file (.so or .dylib file) installed with the HSM's SDK
- The HSM's connector up and running

> ### Configuration

Reviewer: ```suggestion
### 1. Add your HSM information to Notary's configuration file
```

> If you are using the snap you can modify the config under `/var/snap/notary/common/notary.yaml`

Reviewer: I think this line belongs in the config file reference, we don't want to add such a line in every how-to guide that relates to managing the config file.

Author: moved

1. Provide a name to your backend, in the following example we call our backend yubihsm2-backend
2. Add your HSM's information in the config file:
   - Path to the library that is installed with the SDK of your HSM
   - Pin to login on your HSM, this will be in the following format: <auth key id><password>, so if the authentication key used has the id 0001 (default key) your pin might look something like "0001password"
   - ID of the symmetric encryption key that will be used

```yaml
encryption_backend:
  yubihsm2-backend: # name of the backend
    pkcs11:
      lib_path: "/usr/lib/x86_64-linux-gnu/pkcs11/yubihsm_pkcs11.so"
      pin: "0001password"
      aes_encryption_key_id: 0x1234
```

> ### Start Notary

Reviewer: ```suggestion
### 2. Start Notary
```

Author: :white_check_mark:

> If you are using the snap start the service with the following command:
> ```shell
> sudo snap start notary.notaryd
> ```

Reviewer: ```suggestion
Start Notary:
```shell
sudo snap start notary.notaryd
```
```

Author: :white_check_mark:
